launch_ros/launch_ros/utilities/normalize_parameters.py

```diff
@@ -0,0 +1,191 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module with utility for normalizing parameters to a node."""
+
+from collections.abc import Iterable
+from collections.abc import Mapping
+from collections.abc import Sequence
+import pathlib
+from typing import cast
+from typing import List  # noqa
+from typing import Optional
+from typing import Sequence as SequenceTypeHint
+from typing import Tuple  # noqa
+from typing import Union  # noqa
+
+from launch.some_substitutions_type import SomeSubstitutionsType
+from launch.some_substitutions_type import SomeSubstitutionsType_types_tuple
+from launch.substitution import Substitution
+from launch.substitutions import TextSubstitution
+from launch.utilities import ensure_argument_type
+from launch.utilities import normalize_to_list_of_substitutions
+
+from ..parameters_type import ParameterFile
+from ..parameters_type import Parameters
+from ..parameters_type import ParametersDict
+from ..parameters_type import ParameterValue
+from ..parameters_type import SomeParameters
+from ..parameters_type import SomeParametersDict
+from ..parameters_type import SomeParameterValue
+
+
+def _normalize_parameter_array_value(value: SomeParameterValue) -> ParameterValue:
```

Comment:
I think I would need to make it more more specific to make mypy happy `Union[Sequence[str], Sequence[int], Sequence[Substitution]]`. Since the function is prefixed with leading underscore I'll leave it as is for now.

Reply:
Would it be `Sequence[SomeSubstitutions]`? I.e. would you want in the end a single string or a list of strings?

Reply:
> Would it be Sequence[SomeSubstitutions]?  I.e. would you want in the end a single string or a list of strings?

For value or on the return type? For `value`, the types `int` and `float` could passed in via a dictionary in the launch description. For the return type the values need to be given to composable nodes via a ROS service, which means the types have to be kept to know which type to set in the [parameter message](https://github.com/ros2/rcl_interfaces/blob/e631d8f93a48f2aeaa192ecc8ed9bf218330c69a/composition_interfaces/srv/LoadNode.srv#L16). `Sequence[Substitution]` wouldn't work because it [only outputs `Text`](https://github.com/ros2/launch/blob/632ef1f3eca7dbc54fbed85ab65a5efc512d83ed/launch/launch/substitution.py#L37).

Reply:
I was speaking about `value` because I thought that's what @jacobperron was talking about.

I meant as one of the choices in the union you mentioned, not just `Sequence[SomeSubstitutionsType]` (and I do mean that instead of `Sequence[Substitution]`).

I don't have a strong opinion, but I was extrapolating from the function name (contains `array`) and @jacobperron's question. Currently `SomeParameterValue` -> `Union[SomeSubstitutionsType, _SingleValueType, _MultiValueType]`, and `SomeSubstitutionsType` is a list of substitutions, but they're meant to be evaluated and then concatenated, so they'd equate to a "single value type" in the end. If you want to represent a "multi value type" (what I assume is a list of values) then you'd need a list of `SomeSubstitutionsType`.

Maybe there's nothing to be done here, I was just trying to make sure that you're not making the assumption that `SomeSubstitutionsType` ends up being a list of values after evaluations.

Reply:
Ah you're right. Added `Sequence[SomeSubstitutionsType]` (and mypy appeasement required afterwards) to `SomeParameterValue` in 65f23a6cb5b70e4a162af6a6da233dbb851d60c0. The normalized type `ParameterValue` already had the right type for list of lists of substitutions.